### ISSUE TITLE
panic: database is closed

### ISSUE THREAD

**Issue by tomasz1986** [syncthing/syncthing#7244]
I have had this panic today on my Syncthing installation in a VM. I think that this happened when trying to shut down and/or restart Syncthing. I did not even notice it per se when it actually happened, but I just found the panic log file later.

Also, I think this may be related. Before trying to shut down and restart Syncthing, I had removed 1 folder using the Web GUI, but after the restart (and likely the panic), the folder re-appeared, although without `.stfolder`, and hence in an error state.

```
[SWPW3] 12:52:23 WARNING: Fatal error: ghgck-scplg Snapshot(): database is closed
Panic at 2020-12-31T12:52:23+09:00
panic: database is closed

goroutine 2162 [running]:
github.com/syncthing/syncthing/lib/db.obfuscateAndPanic(0x110edc0, 0x1690dc8)
	github.com/syncthing/syncthing/lib/db/lowlevel.go:1172 +0x6f
github.com/syncthing/syncthing/lib/db.fatalError(0x110edc0, 0x1690dc8, 0x157220e0, 0x16, 0x1208e400)
	github.com/syncthing/syncthing/lib/db/set.go:535 +0x113
github.com/syncthing/syncthing/lib/db.(*FileSet).Snapshot(0x12c939e0, 0x1237f320)
	github.com/syncthing/syncthing/lib/db/set.go:142 +0x235
github.com/syncthing/syncthing/lib/model.(*model).Availability(0x12085c00, 0x12382ef4, 0xb, 0x1612ca00, 0x45, 0x1500285, 0x0, 0x5fec9631, 0x0, 0xf018a7f2, ...)
	github.com/syncthing/syncthing/lib/model/model.go:2392 +0x18a
github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullBlock(0x1241b680, 0x12180540, 0x16121780, 0x20, 0x20, 0x2e0000, 0x0, 0x20000, 0x684f8793, 0x12bcf4c0)
	github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:1491 +0xce
github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerRoutine.func1(0x11176a0, 0x15152450, 0x12390a80, 0x20000, 0x1241b680, 0x12180540, 0x16121780, 0x20, 0x20, 0x2e0000, ...)
	github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:1466 +0x8d
created by github.com/syncthing/syncthing/lib/model.(*sendReceiveFolder).pullerRoutine
	github.com/syncthing/syncthing/lib/model/folder_sendrecv.go:1462 +0x268
```

[panic-20201231-125223.log](https://github.com/syncthing/syncthing/files/5756418/panic-20201231-125223.log)

Syncthing version is the same as in https://github.com/syncthing/syncthing/issues/7231, i.e. v1.12.0 with 5 custom commits.

**Comment by imsodin** [comment]
That's the same root cause as #7005 (returning nil filesets on shutdown), and that was fixed by #7150.

**Comment by tomasz1986** [comment]
Thank you. I am sorry for the log spam then. I would like to be able to determine myself whether this is the same issue or not, but the logs always seem to look quite a bit differently each time. I am seeing that the commit is scheduled for 1.13.0, which means still quite a while to go, so I guess that I will need to do some cherry picking in order to use it early.

**Comment by imsodin** [comment]
Nothing to be sorry about, this is an ugly issue and can happen under various circumstances. I don't want to deter anyone from reporting panics, I am just happily able to point at a solution here and thus closing ;)  
For context: Basically all "db is closed" panics and nil pointer derefs on `db.FileSet` members should be fixed by #7150. However these should not occur consistently, but only if you stop Syncthing while it is still in the process of (re-)starting folders (Syncthing startup or config changes) - meaning it's only crashing when it's already shutting down (I am not intending to justify/play anything down, crashing is never ok).
